cairo1-run/src/cairo_run.rs

```diff
@@ -483,98 +504,122 @@ fn create_entry_code(
+        BuiltinName::poseidon => builtin_vars[&PoseidonType::ID],
+        BuiltinName::segment_arena => builtin_vars[&SegmentArenaType::ID],
+        BuiltinName::keccak
+        | BuiltinName::ecdsa
+        | BuiltinName::range_check96
+        | BuiltinName::add_mod
+        | BuiltinName::mul_mod => unreachable!(),
+    };
+    if copy_to_output_builtin && got_segment_arena {
+        for (i, builtin) in builtins.iter().enumerate() {
+            let var = get_var(builtin);
+            let local = ctx.add_var(CellExpression::Deref(deref!([fp + i.to_i16().unwrap()])));
+            casm_build_extend!(ctx, assert local = var;);
+        }
+        let segment_arena_ptr = get_var(&BuiltinName::segment_arena);
+        casm_build_extend! {ctx,
+            tempvar n_segments = segment_arena_ptr[-2];
+            tempvar n_finalized = segment_arena_ptr[-1];
+            assert n_segments = n_finalized;
+            jump STILL_LEFT_PRE if n_segments != 0;
+            rescope{};
+            jump DONE_VALIDATION;
+            STILL_LEFT_PRE:
+            const one = 1;
+            tempvar infos = segment_arena_ptr[-3];
+            tempvar remaining_segments = n_segments - one;
+            rescope{infos = infos, remaining_segments = remaining_segments};
+            LOOP_START:
+            jump STILL_LEFT_LOOP if remaining_segments != 0;
+            rescope{};
+            jump DONE_VALIDATION;
+            STILL_LEFT_LOOP:
+            const one = 1;
+            const three = 3;
+            tempvar prev_end = infos[1];
+            tempvar curr_start = infos[3];
+            assert curr_start = prev_end + one;
+            tempvar next_infos = infos + three;
+            tempvar next_remaining_segments = remaining_segments - one;
+            rescope{infos = next_infos, remaining_segments = next_remaining_segments};
+            #{ steps = 0; }
+            jump LOOP_START;
+            DONE_VALIDATION:
+        };
+        for i in 0..builtins.len().to_i16().unwrap() {
+            let local = ctx.add_var(CellExpression::Deref(deref!([fp + i])));
+            casm_build_extend!(ctx, tempvar _r = local;);
+        }
+    } else {
+        for builtin in &builtins {
+            let var = get_var(builtin);
+            casm_build_extend!(ctx, tempvar _r = var;);
         }
-        ret_builtin_offset += 1;
     }
 
     if config.proof_mode {
-        casm_extend!(ctx, jmp rel 0;);
+        casm_build_extend! {ctx,
+            INFINITE_LOOP:
+            // To enable the merge of the branches.
+            #{ steps = 0; }
+            jump INFINITE_LOOP;
+        };
     } else {
-        casm_extend!(ctx, ret;);
+        casm_build_extend!(ctx, ret;);
     }
-
-    assert_eq!(before_final_call + final_call_size, ctx.current_code_offset);
-
-    Ok((ctx, builtins))
+    let result = ctx.build(["FUNCTION"]);
+    let [call_inst] = result.branches[0].1.as_slice() else {
+        panic!("Expected a single relocation");
```

Comment:
Can we avoid the panics here? If it is to much work just create an issue and we will handle it

Reply:
this cannot ever actually panic - given that the builder code is correct (it will either always crash if we've written bad code here) or never crash (if the code here is fine - which it is)